configfile: "config/config.yaml"
include: "rules/common.py"

include: "rules/DownloadBAM.smk"
include: "rules/GetGeneAnnotation.smk"
#include: "rules/GetCountTables.smk"
include: "rules/Develop.smk"
#include: "rules/ExpressionAnalysis.smk"
#include: "rules/RunTests.smk"
include: "rules/DifferentialSplicing.smk"
include: "rules/ebpmf.smk"
include: "rules/GetCoverageAndCounts.smk"
include: "rules/RunRMATS.smk"
include: "rules/RunSleuth.smk"
include: "rules/ProcessGREGoR.smk"
include: "rules/isoforms.smk"
include: "rules/QTLs.smk"

localrules:
    all,
    DownloadFromGTEx_Brain_Cortex,
    DownloadFromGTEx_Muscle_Skeletal,
    DownloadFromGTEx_Whole_Blood,
    DownloadFromGTEx_Liver,
    DownloadFromGTEx_Brain_Hippocampus,
    DownloadFromGTEx_Brain_Hypothalamus,
    DownloadHg38Ref,
    DownloadFromGTEx_TestSamples,
    DownloadFromGTEx_TestSamples_juncs,
    DownloadTranscriptome


rule all:
    input:
        #expand("stm_models/{gene}/{gene}.sgom_K{knum}.rds", 
        #       gene=genes,
        #       knum=[3, 5, 10]
        #),
        expand("stm_models/{gene}/{gene}.sgom_K{knum}.rds", 
          gene=["CERCAM", "IDH3B", "GRHPR", "SNAP25", "FLNA", "RPN2", "NDRG4", "MORF4L2", "CMC1",
                "PKM", "NPM3", "FGF8", "CMC2", "APOE", "CKS2"], 
          knum=[3, 5, 10, 2]
        ),        
        #expand("stm_models/{gene}/{gene}.{mode}_{tissue}.sgom_K{knum}.rds",
        #        mode = ["no", "one_of", "minus_one_of"],
        #        gene = ["TPM3", "SRSF3"], 
        #        tissue = ["Muscle_Skeletal", "Whole_Blood", "Brain_Cortex"],
        #        knum=[3, 5, 10]
        #        )
        

        
        
rule test_ebpmf_alpha:
    input:
        expand("coverage/counts_filtered/{gene}.csv.gz", gene = prueba_genes + list(selected_genes.gene)),
        

# awk '$20 <= 0.1 && $23>=0.1 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - > config/genes.Brain_Cortex_v_Muscle_Sleketal.all_positives.txt

# awk '$20 <= 0.1 && $23<=-0.1 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - >> config/genes.Brain_Cortex_v_Muscle_Sleketal.all_positives.txt

# awk '$20 >= 0.25 && $23<=0.05 && $23>=-0.05 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - | grep -v ENSG - > config/genes.Brain_Cortex_v_Muscle_Sleketal.negative_candidates.txt

# awk 'FNR == NR { h[$1,$2]; next }; !($1 SUBSEP $2 in h)' config/genes.Brain_Cortex_v_Muscle_Sleketal.all_positives.txt config/genes.Brain_Cortex_v_Muscle_Sleketal.negative_candidates.txt > config/genes.Brain_Cortex_v_Muscle_Sleketal.all_negatives.txt


# shuf -n 200 config/genes.Brain_Cortex_v_Muscle_Sleketal.all_negatives.txt > config/genes.Brain_Cortex_v_Muscle_Sleketal.negatives.txt


# awk '$20 <= 0.1 && $23<=0.1 && $23>=-0.1 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - | shuf -n 200 > config/genes.Brain_Cortex_v_Muscle_Sleketal.negatives.txt

#  awk '$20 <= 0.1 && $23>=0.1 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - | shuf -n 100 > config/genes.Brain_Cortex_v_Muscle_Sleketal.positives.txt

# awk '$20 <= 0.1 && $23<=-0.1 ' DifferentialSplicing/rMATS/Brain_Cortex_v_Muscle_Skeletal_train/output/SE.MATS.JCEC.txt | awk -F'"' '{print $4}' - | sort -u | grep -v '.1' - | shuf -n 100 >> config/genes.Brain_Cortex_v_Muscle_Sleketal.positives.txt
